File: 09/09-lookup/singleBit/RTable.c
#include "RTable.h"
#include <stdbool.h>
static inline int getIndexFromIP(uint32_t IP){
    return IP >> 31;
}

void initNode(Node *node){
    node->nodeKind = Internal;
    node->port = -1;
    node->children[LEFT] = NULL;
    node->children[RIGHT] = NULL;
}

void initTree(RTableTree *tree) {
    initNode(tree);
}

Node *createNewNode() {
    Node *returnNode = (Node*)malloc(sizeof(Node));
    if (returnNode != NULL) {
        initNode(returnNode);
        return returnNode;
    } else {
        fprintf(stdout, "No memory\n");
    }
    return NULL;
}

// set bitfileds of a node
void setNode(Node* node, PortID port, NodeKind nodeKind) {
    node->nodeKind = nodeKind;
    node->port = port;
}

// add a new IP to the correct node of a tree
void addNewIP(RTableTree *tree, uint32_t IP, int prefixLength, PortID port) {
    int index = getIndexFromIP(IP);
    if (tree->children[index] == NULL) {
        tree->children[index] = createNewNode();
        if (tree->children[index] == NULL) {
            fprintf(stdout, "No memory\n");
            exit(-1);
        }
    }
    if (prefixLength == 1) {
        setNode(tree->children[index], port, Matched);
    } else {
        addNewIP(tree->children[index], IP << 1, prefixLength - 1, port);
    }
}

static bool isLeaf(RTableTree *tree) {
    return tree->children[LEFT] == NULL && tree->children[RIGHT] == NULL;
}

static void matchIPHelper(RTableTree *tree, uint32_t IP, PortID *portPtr){
    if (!tree)
        return;
    if (tree->nodeKind == Matched)
        *portPtr = tree->port;
    
    if (isLeaf(tree)) {
        return;
    } else {
        matchIPHelper(tree->children[getIndexFromIP(IP)], IP << 1, portPtr);
    }
}

PortID matchIP(RTableTree *tree, uint32_t IP) {
    PortID portID = -1;
    matchIPHelper(tree, IP, &portID);
    return portID;
}

// recycle all space used by the tree
void destroyTree(RTableTree * tree) {
    if (tree->children[LEFT] != NULL)
        destroyTree(tree->children[LEFT]);

    if (tree->children[RIGHT] != NULL)
        destroyTree(tree->children[RIGHT]);

    free(tree);
    tree = NULL;
}

File: 09/09-lookup/singleBit/RTable.go
package ratbletrie

// Node tree node type
type Node struct {
	nodeKind string
	Port     int
	children [2]*Node
}

// New return a new node
func New() *Node {
	var node Node
	node.InitNode()
	return &node
}

// InitNode initializa a node
func (node *Node) InitNode() {
	node.children[0] = nil
	node.children[1] = nil
	node.Port = -1
	node.nodeKind = "internal"
}

// SetNode sets a node's attributes
func (node *Node) SetNode(port int, kind string) {
	node.Port = port
	node.nodeKind = kind
}

func getIndexFromIP(IP uint32) int {
	return int(IP >> 31)
}

// AddNewIP add new node to a trie
func (node *Node) AddNewIP(IP uint32, prefixLength int, port int) {
	index := getIndexFromIP(IP)
	if node.children[index] == nil {
		node.children[index] = New()
	}

	if prefixLength == 1 {
		node.children[index].SetNode(port, "matched")
	} else {
		node.children[index].AddNewIP(IP<<1, prefixLength-1, port)
	}
}
